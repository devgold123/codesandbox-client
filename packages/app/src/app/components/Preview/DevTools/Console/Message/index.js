import React from 'react';

import { ObjectInspector } from 'react-inspector';

import formatMessageString from '../format-message';
import MessageIcon from '../MessageIcon';
import { IconContainer } from '../elements';
import { Container, InnerItem, inspectorTheme } from './elements';

function formatMessage(message) {
  const formattedResult = document.createElement('span');

  formatMessageString(
    message.arguments[0],
    message.arguments.slice(1),
    formattedResult
  );

  return formattedResult;
}

function getMessage(message) {
  if (message.type === 'return') {
    return (
      <InnerItem>
        <ObjectInspector theme={inspectorTheme} data={message.arguments[0]} />
      </InnerItem>
    );
  }

  if (message.type === 'command') {
    return <InnerItem>{message.arguments[0]}</InnerItem>;
  }

  if (
    message.arguments.length > 0 &&
    typeof message.arguments[0] === 'string' &&
    message.arguments[0].indexOf('%') > -1
  ) {
    return (
      <InnerItem
        dangerouslySetInnerHTML={{
          __html: formatMessage(message).outerHTML.replace(
            /(?:\r\n|\r|\n)/g,
            '<br />'
          ),
        }}
      />
    );
  }

  if (message.arguments.every(argument => typeof argument === 'string')) {
    return <InnerItem>{message.arguments.join(' ')}</InnerItem>;
  }

  return message.arguments.map((m, i) => (
    // eslint-disable-next-line react/no-array-index-key
    <InnerItem key={i}>
      <ObjectInspector theme={inspectorTheme} data={m} />
    </InnerItem>
  ));
}

function Message({ message }) {
  return (
    <Container logType={message.logType}>
      <div>
        <IconContainer>
          <MessageIcon type={message.type} logType={message.logType} />
        </IconContainer>
      </div>
      <div>{getMessage(message)}</div>
    </Container>
  );
}

export default Message;
