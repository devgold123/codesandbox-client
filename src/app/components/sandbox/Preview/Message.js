import React from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import styled from 'styled-components';
import InfoIcon from 'react-icons/lib/md/info';
import WarningIcon from 'react-icons/lib/md/warning';
import ErrorIcon from 'react-icons/lib/md/error';

import Button from 'app/components/buttons/Button';
import sandboxActionCreators from 'app/store/entities/sandboxes/actions';

import Editor from './Editor';

const Container = styled.div`
  font-family: 'Source Code Pro', monospace;
  font-weight: 400;
  display: flex;
  justify-content: center;
  padding: 0.5rem;
  background-color: ${({ color, theme }) => {
  if (color === 'error') {
    return theme.redBackground;
  }
  return theme.primary;
}}
  color: ${props => props.theme.red};
  height: 100%;
  width: 100%;
  padding: 1rem;
  vertical-align: middle;
  line-height: 1.4;
  text-align: center;
`;

const Content = styled.div`
  width: 70%;
`;

const Icon = styled.div`
  margin: 1rem;
  font-size: 3rem;
`;

type Props = {
  sandboxId: string,
  error: ?{ type: string, title: string, message: string },
  message: ?string,
  sandboxActions: typeof sandboxActionCreators,
};

const mapDispatchToProps = dispatch => ({
  sandboxActions: bindActionCreators(sandboxActionCreators, dispatch),
});
class Message extends React.PureComponent {
  props: Props;

  getIcon = () => {
    const { error, message } = this.props;

    if (message) return <Icon><InfoIcon /></Icon>;
    if (error.severity === 'warning') return <Icon><WarningIcon /></Icon>;

    return <Icon><ErrorIcon /></Icon>;
  };

  addDependency = name => async () => {
    const { sandboxId, sandboxActions } = this.props;
    await sandboxActions.addNPMDependency(sandboxId, name);
    sandboxActions.fetchDependenciesBundle(sandboxId);
  };

  openModule = () => {
    const { error, sandboxId, sandboxActions } = this.props;
    sandboxActions.setCurrentModule(sandboxId, error.moduleId);
  };

  getMessage = () => {
    const { error, modules } = this.props;

    if (error.type === 'dependency-not-found') {
      return (
        <div>
          Could not find the dependency
          <b> {"'"}{error.payload.dependency}{"'"}</b>.
          <div style={{ marginTop: '1rem' }}>
            <Button
              onClick={this.addDependency(error.payload.dependency)}
              small
            >
              ADD DEPENDENCY TO PROJECT
            </Button>
          </div>
        </div>
      );
    } else if (error.type === 'no-dom-change') {
      return (
        <div>
          It seems like <b>{module.title}</b> doesn{"'"}t render anything.
          <div style={{ marginTop: '1rem' }}>
            We
            {"'"}
            re adding support for viewing generated documentation in the near future.
            {error.payload.react &&
              <div>
                For now we recommend adding
                {' '}
                <Editor readOnly name={error.payload.componentName} />
                to view this component.
              </div>}
          </div>
        </div>
      );
    }

    return (
      <div>
        {module && <div>Error in <b>{module.title}</b>:</div>}
        <div style={{ marginTop: '1rem' }}>{error.title}: {error.message}</div>

        {module &&
          <div style={{ marginTop: '1rem' }}>
            <Button onClick={this.openModule} small>
              OPEN MODULE
            </Button>
          </div>}
      </div>
    );
  };

  render() {
    const { error, message } = this.props;

    return (
      <Container color={error && error.severity === 'error' && 'error'}>
        <Content>
          {this.getIcon()}
          {message ? <div>{message}</div> : this.getMessage()}
        </Content>
      </Container>
    );
  }
}

export default Message;
