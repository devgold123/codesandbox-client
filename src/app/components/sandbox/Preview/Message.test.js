import React from 'react';
import renderer from 'react-test-renderer';
import Message from './Message';

describe('Message', () => {
  test('it renders a dependency not found error', () => {
    const tree = renderer
      .create(
        <Message
          sandboxId="foeijf"
          error={{
            type: 'dependency-not-found',
            title: 'Dependency Not Found',
            message: 'Dependendnen',
            severity: 'error',
            moduleId: 'asfa',
            payload: {
              path: 'react-icons/fa/beer',
              dependency: 'react-icons',
            },
          }}
          modules={[
            {
              id: 'asfa',
              code: 'daowidj',
            },
          ]}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('it renders a warning', () => {
    const tree = renderer
      .create(
        <Message
          sandboxId="foeijf"
          error={{
            severity: 'warning',
            title: 'Loading dependencies',
          }}
          modules={[
            {
              id: 'asfa',
              code: 'daowidj',
            },
          ]}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
