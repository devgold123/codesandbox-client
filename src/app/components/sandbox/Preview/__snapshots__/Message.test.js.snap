// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Message it renders a dependency not found error 1`] = `
<div
  className="sc-ifAKCX jHZoMf"
  color="error"
>
  <div
    className="sc-EHOje gTncrN"
  >
    <div
      className="sc-bZQynM kFeJzc"
    >
      <svg
        fill="currentColor"
        height="1em"
        preserveAspectRatio="xMidYMid meet"
        style={
          Object {
            "color": undefined,
            "verticalAlign": "middle",
          }
        }
        viewBox="0 0 40 40"
        width="1em"
      >
        <g>
          <path
            d="m21.6 21.6v-10h-3.2v10h3.2z m0 6.8v-3.4h-3.2v3.4h3.2z m-1.6-25c9.2 0 16.6 7.4 16.6 16.6s-7.4 16.6-16.6 16.6-16.6-7.4-16.6-16.6 7.4-16.6 16.6-16.6z"
          />
        </g>
      </svg>
    </div>
    <div>
      Could not find
      <b>
         
        '
        react-icons/fa/beer
        '
      </b>
      .
      <div>
        Did you add the dependency
         
        <b>
          '
          react-icons
          '
        </b>
        ?
      </div>
    </div>
  </div>
</div>
`;

exports[`Message it renders a warning 1`] = `
<div
  className="sc-ifAKCX fxFwBM"
  color={false}
>
  <div
    className="sc-EHOje gTncrN"
  >
    <div
      className="sc-bZQynM kFeJzc"
    >
      <svg
        fill="currentColor"
        height="1em"
        preserveAspectRatio="xMidYMid meet"
        style={
          Object {
            "color": undefined,
            "verticalAlign": "middle",
          }
        }
        viewBox="0 0 40 40"
        width="1em"
      >
        <g>
          <path
            d="m21.6 23.4v-6.8h-3.2v6.8h3.2z m0 6.6v-3.4h-3.2v3.4h3.2z m-20 5l18.4-31.6 18.4 31.6h-36.8z"
          />
        </g>
      </svg>
    </div>
    <div>
      <div>
        Error in 
        <b />
        :
      </div>
      <div
        style={
          Object {
            "marginTop": "1rem",
          }
        }
      >
        Loading dependencies
        : 
      </div>
    </div>
  </div>
</div>
`;
